chalk-solve/src/rust_ir.rs

```diff
@@ -138,6 +138,24 @@ pub struct FnDefDatumBound<I: Interner> {
     ///                ^^^
     /// ```
     pub return_type: Ty<I>,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash, Fold, HasInterner)]
+/// Represents the bounds on a `FnDefDatum`, including
+/// the function definition's type signature and where clauses.
+pub struct FnDefDatumBound<I: Interner> {
+    /// Inputs and outputs defined on a function
+    /// These are needed for late-bound regions in rustc. For example the
+    /// lifetime `'a` in
+    /// ```ignore
+    /// fn foo<'a, T>(&'a T);
+    ///        ^^
+    /// ```
+    /// Rustc doesn't pass in late-bound the regions in substs, but the inputs
+    /// and outputs may use them. `where_clauses` don't need an extra set of
+    /// `Binders`, since any lifetimes found in where clauses are not late-bound.
```

Comment:
```suggestion
    /// `Binders`, since any lifetimes found in where clauses are not late-bound.
    ///
    /// For more information, see [this rustc-dev-guide chapter](https://rustc-dev-guide.rust-lang.org/early-late-bound.html).
```